File: frontend/src/lib/attribution.ts
/**
 * Where a visitor came from, remembered until they order.
 *
 * Nothing in the platform recorded this before: a brand running ads could not
 * tell which campaign paid for an order, because the `utm_*` values on the
 * landing URL were read by nobody.
 *
 * Two rules decide what is kept:
 *
 * **Last non-direct wins.** Somebody clicks a Google ad, browses, leaves, and
 * comes back a day later by typing the address. The ad still sold it, so the
 * second visit — which carries no campaign — must not erase the first. Only a
 * visit that names a source replaces the stored one.
 *
 * **An unknown source stays unknown.** A visitor who arrived with no campaign
 * and no referrer gets no attribution at all. Writing "direct" would turn
 * something we do not know into something the brand would act on.
 *
 * Kept in localStorage, which is per-origin — and every brand has its own
 * subdomain, so one brand's campaigns can never be read on another's store.
 * Every access is wrapped: private browsing makes these throw.
 */

const KEY = "at_attr";
const MAX_AGE_DAYS = 90;
/** A gap this long counts as a new visit rather than more of the same one. */
const VISIT_GAP_MINUTES = 30;

const UTM_KEYS = ["utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content"] as const;
const CLICK_KEYS = ["gclid", "fbclid", "msclkid", "ttclid"] as const;

export interface Attribution {
  utm_source?: string;
  utm_medium?: string;
  utm_campaign?: string;
  utm_term?: string;
  utm_content?: string;
  gclid?: string;
  fbclid?: string;
  msclkid?: string;
  ttclid?: string;
  referrer?: string;
  landing_page?: string;
  first_seen?: string;
  last_seen?: string;
  visits?: number;
  device?: "mobile" | "tablet" | "desktop";
}

function read(): Attribution | null {
  try {
    const raw = window.localStorage.getItem(KEY);
    if (!raw) return null;
    const stored = JSON.parse(raw) as Attribution;
    // Expire rather than credit an order to a campaign from months ago.
    if (stored.first_seen) {
      const age = Date.now() - new Date(stored.first_seen).getTime();
      if (age > MAX_AGE_DAYS * 86_400_000) return null;
    }
    return stored;
  } catch {
    return null;
  }
}

function write(value: Attribution): void {
  try {
    window.localStorage.setItem(KEY, JSON.stringify(value));
  } catch {
    /* private browsing, or storage full — attribution is never worth an error */
  }
}

function deviceKind(): Attribution["device"] {
  const ua = navigator.userAgent;
  if (/iPad|Tablet|PlayBook|Silk|(Android(?!.*Mobile))/i.test(ua)) return "tablet";
  if (/Mobi|Android|iPhone|iPod|Windows Phone/i.test(ua)) return "mobile";
  return "desktop";
}

/** The referrer, but only when it came from somewhere else. */
function externalReferrer(): string | undefined {
  const ref = document.referrer;
  if (!ref) return undefined;
  try {
    if (new URL(ref).host === window.location.host) return undefined;
  } catch {
    return undefined;
  }
  return ref.slice(0, 1000);
}

/**
 * Record this page view. Safe to call on every navigation.
 *
 * Returns what is now stored, so a caller can act on it without a second read.
 */
export function captureAttribution(): Attribution | null {
  if (typeof window === "undefined") return null;

  const params = new URLSearchParams(window.location.search);
  const incoming: Attribution = {};
  for (const key of UTM_KEYS) {
    const value = params.get(key)?.trim();
    if (value) incoming[key] = value.slice(0, 255);
  }
  for (const key of CLICK_KEYS) {
    const value = params.get(key)?.trim();
    if (value) incoming[key] = value.slice(0, 255);
  }

  const referrer = externalReferrer();
  // A campaign, an ad click, or a link from another site: all of them name
  // where this visit came from, and any of them outranks what we had.
  const namesASource = Object.keys(incoming).length > 0 || !!referrer;

  const now = new Date().toISOString();
  const stored = read();

  if (!stored) {
    if (!namesASource) return null; // nothing to say, so say nothing
    const fresh: Attribution = {
      ...incoming,
      referrer,
      landing_page: (window.location.origin + window.location.pathname).slice(0, 1000),
      first_seen: now,
      last_seen: now,
      visits: 1,
      device: deviceKind(),
    };
    write(fresh);
    return fresh;
  }

  const gapMs = stored.last_seen ? Date.now() - new Date(stored.last_seen).getTime() : Infinity;
  const newVisit = gapMs > VISIT_GAP_MINUTES * 60_000;

  const next: Attribution = { ...stored, last_seen: now };
  if (newVisit) next.visits = (stored.visits ?? 1) + 1;

  if (namesASource) {
    // A fresh source replaces the old one outright: keeping half of the
    // previous campaign beside half of this one would describe no real visit.
    for (const key of [...UTM_KEYS, ...CLICK_KEYS]) delete next[key];
    Object.assign(next, incoming);
    next.referrer = referrer ?? undefined;
  }

  write(next);
  return next;
}

/** What to send with the order, or null when nothing was ever recorded. */
export function getAttribution(): Attribution | null {
  if (typeof window === "undefined") return null;
  const stored = read();
  if (!stored) return null;
  // Drop empty keys so a mostly-blank record does not read as a real one.
  const payload = Object.fromEntries(
    Object.entries(stored).filter(([, v]) => v !== undefined && v !== null && v !== "")
  ) as Attribution;
  return Object.keys(payload).length ? payload : null;
}

/** Forget the visitor's attribution — used after an order is placed. */
export function clearAttribution(): void {
  try {
    window.localStorage.removeItem(KEY);
  } catch {
    /* nothing to do */
  }
}

File: backend/app/services/attribution.py
"""Where an order came from.

The browser collects this (see frontend/src/lib/attribution.ts) and sends it
once, at checkout. Everything here treats that payload as untrusted: it arrives
from a URL anybody can edit, so every value is trimmed, length-capped and
dropped if it is not a string.

Two rules shape the rest:

* An unknown source is recorded as unknown. A buyer who typed the address in
  has no campaign, and filling the blank with "direct" would turn a fact we do
  not have into one a brand would go on to act on.
* Last non-direct wins. If someone arrives from a Google ad, leaves, and comes
  back by typing the address, the ad is still what sold it — the second visit
  carries no campaign, so it must not erase the first. The browser decides
  this; the server only stores the answer.
"""
from __future__ import annotations

from typing import Any

# The five a brand filters, groups and exports by. These get real columns.
UTM_FIELDS = ("utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content")

# Everything else about the visit, kept together in one JSONB column.
META_FIELDS = (
    "referrer",       # the page that linked here
    "landing_page",   # the first page of this store they saw
    "gclid",          # Google Ads click
    "fbclid",         # Meta click
    "msclkid",        # Microsoft Ads click
    "ttclid",         # TikTok click
    "first_seen",     # ISO time of their first visit
    "last_seen",      # ISO time of the visit that placed the order
    "visits",         # how many visits before ordering
    "device",         # mobile | tablet | desktop
)

_MAX = 255
_MAX_URL = 1000


def _text(value: Any, limit: int = _MAX) -> str | None:
    if not isinstance(value, str):
        return None
    cleaned = value.strip()
    return cleaned[:limit] if cleaned else None


def clean(payload: Any) -> dict[str, Any]:
    """A safe, normalised copy of what the browser sent.

    Returns `{"utm": {...}, "meta": {...}}`, both possibly empty. Anything not
    recognised is discarded rather than stored, so a crafted URL cannot push
    arbitrary data into the orders table.
    """
    if not isinstance(payload, dict):
        return {"utm": {}, "meta": {}}

    utm = {
        field: value
        for field in UTM_FIELDS
        # utm_source is lower-cased so "Google", "google" and "GOOGLE" group as
        # one campaign source rather than three.
        if (value := _text(payload.get(field))) is not None
    }
    if "utm_source" in utm:
        utm["utm_source"] = utm["utm_source"].lower()
    if "utm_medium" in utm:
        utm["utm_medium"] = utm["utm_medium"].lower()

    meta: dict[str, Any] = {}
    for field in META_FIELDS:
        raw = payload.get(field)
        if field == "visits":
            try:
                count = int(raw)
            except (TypeError, ValueError):
                continue
            if 0 < count < 10_000:
                meta[field] = count
            continue
        limit = _MAX_URL if field in ("referrer", "landing_page") else _MAX
        value = _text(raw, limit)
        if value is not None:
            meta[field] = value

    return {"utm": utm, "meta": meta}


def apply(order: Any, payload: Any) -> bool:
    """Put a cleaned payload on an order. True if anything was recorded.

    Never overwrites attribution an order already has: it is a record of how
    that one order arrived, and a later edit cannot know better than the visit
    that placed it.
    """
    data = clean(payload)
    if not data["utm"] and not data["meta"]:
        return False
    if any(getattr(order, field, None) for field in UTM_FIELDS) or getattr(order, "attribution", None):
        return False

    for field, value in data["utm"].items():
        setattr(order, field, value)
    if data["meta"]:
        order.attribution = data["meta"]
    return True


def summary(order: Any) -> dict[str, Any] | None:
    """How this order arrived, ready for the order page — or None if unknown.

    `label` is the one line a brand reads at a glance: "google / cpc" and, when
    there is one, the campaign that paid for it.
    """
    utm = {field: getattr(order, field, None) for field in UTM_FIELDS}
    meta = getattr(order, "attribution", None) or {}
    if not any(utm.values()) and not meta:
        return None

    source = utm.get("utm_source")
    medium = utm.get("utm_medium")
    if source:
        label = f"{source} / {medium}" if medium else source
    elif meta.get("referrer"):
        label = f"Referral — {meta['referrer']}"
    else:
        # Something was recorded (a click id, a landing page) but nothing that
        # names a source. Say so rather than guessing at one.
        label = "Unattributed"

    return {
        **{field: value for field, value in utm.items() if value},
        "label": label,
        "campaign": utm.get("utm_campaign"),
        "referrer": meta.get("referrer"),
        "landing_page": meta.get("landing_page"),
        "visits": meta.get("visits"),
        "device": meta.get("device"),
        "first_seen": meta.get("first_seen"),
        "paid_click": next(
            (name for name in ("gclid", "fbclid", "msclkid", "ttclid") if meta.get(name)), None
        ),
    }
